### ISSUE TITLE
Lua: Allow display of listchars other than eol

### ISSUE THREAD

**Issue by timbedard** [lukas-reineke/indent-blankline.nvim#74]
indentLine would leave listchars like `tab` displayed, rather than replace them with its lines as indent-blankline is doing right now. With the `tab` listchar set, I think showing it is the expected behavior.

Similarly, when a line contains only spaces and `trail` is set, I would expect the list character to be displayed instead of the indent lines.

**Comment by lukas-reineke** [comment]
both sound like good options 👍 
I'll look into it

**Comment by WhyNotHugo** [comment]
I've managed to bisect the issue, and the regression start in d9f682939ac8a59b1784802b0c69e82177a4972a.

However, I'm not proficient enough in lua to further pinpoint what's up there.

**Comment by lukas-reineke** [comment]
It's not really a regression. The whole functionality of the plugin changed.
Before using the new overlay virtual text, the plugin only did empty lines. And by definition, those didn't have trailing spaces.

**Comment by lcrockett** [comment]
Personally i'd prefer to keep being able to see the difference between tabs and spaces using the `listchars` defined characters.

Would it be possible to accomplish this in any way ?

**Comment by ek234** [comment]
hey,
I am a comp sci student. I wanted to start open-source contributions.

Can I work on this issue/feature request?

**Comment by lukas-reineke** [comment]
Hey @ek234,
sure! Always grateful for help.
What I am thinking for this feature is an option that is a list of all listchars that should take priority over indent-blankline virtual text.
So for example `let g:indent_blankline_foo = ["eol", "tab"]` would not overwrite tab characters and end-of-line characters.

I think we need to care about `eol`, `tab`, `space`, `trail` and `lead`, but I haven't looked into it too much.

It will get complicated when there are cases where a line has a mix of showing indentation lines and not showing them. Like mixed indentation with spaces and tabs. It might make sense to just not support this for now, not sure.

If you start working on this, open a draft PR early. I'll review the code and help you, so you don't spend a lot of time on something I won't want to merge.

**Comment by ek234** [comment]
Hi @lukas-reineke,

Even I thought of implementing it with a similar approach. However, an issue came up.

Consider in listchar that tab is set to anything other than '|\ ', for example, let it be "tab:>-" [like in the nvim documentation].
Then all the non blank lines (which actually contain tabs for indenting) will use this scheme and the plugin wouldn't write over it. But, the blank lines will still use the regular hard coded scheme.

>--->---this is what it would look like
|     |     |
>--->---the line just above this line was blank
>--->---more text
>--->---more text
 
Rather, I would suggest that plugin itself displays tabs as the user wishes. This way, the indent will be uniform.
Additionally, we don't have to worry about which char takes priority over the indent line.

Also, as eol is already considered in the plugin, it works perfectly for me.
I believe other listchars can be inserted in a way similar to eol.

tl;dr We can have the plugin display all the characters as defined by the user instead of checking if the underlying char is to be prioritised or not. This will increase customisability too.

**Comment by ek234** [comment]
I still feel displaying the list chars in the virtual text will be more consistent.

One way of implementing it would be kinda similar to list char in nvim itself. we can store the pattern that the user wants. for eg, for '|.|', we can just split the tab display into these three parts: display '|' once then display '.' as many times as needed then display '|'.

The problem with not displaying virtual text over list char is that we will still have to display the same tab list chars for the blank lines. (as I mentioned in the last message). so we'll have to implement both the things.

Please let me know if you're fine with this approach.
Thanks

**Comment by lukas-reineke** [comment]
> I still feel displaying the list chars in the virtual text will be more consistent.

Like I said, that is one of the 2 options. I am fine with this, whatever is easier to implement.

> we can store the pattern that the user wants

This I don't agree with. Read the value from listchars directly. No custom pattern to store anything.

**Comment by ek234** [comment]
> Like I said, that is one of the 2 options. I am fine with this, whatever is easier to implement.
ahh cool:)
> This I don't agree with. Read the value from listchars directly. No custom pattern to store anything.
ohh that would be nice. I didn't think of that at all.

**Comment by tsakirist** [comment]
Any updates on this? Not being able to see if a file's indentations contains tabs characters is a real pain.

**Comment by timbedard** [comment]
I haven't really had time to take a stab at it, myself. I imagine @lukas-reineke hasn't had a chance yet either.

**Comment by lukas-reineke** [comment]
It's a pretty big task. I am not sure when I have time for this.

PR's are always welcome. I will do my best to help if someone wants to try this again.

**Comment by ryuheechul** [comment]
I was wondering why my tab char wasn't shown so I had fought with neovim for an hour and finally arrived to the right place 😅.
(digging of it allow me to learn more about `list` and `listchars` though!)

I think I'm going to stop loading `ident-blankline.nvim` until this issue gets resolved. All the good luck with the task!

**Comment by Daxtorim** [comment]
Not sure if just mentioning issues sends out notifications, so I'll just quickly leave a comment here. I took a stab at this in https://github.com/lukas-reineke/indent-blankline.nvim/pull/187
Some testing/feedback would be appreciated.

**Comment by timbedard** [comment]
🎉
